### PR TITLE
#1402 Accessibility: Fixed Tooltip element's role 'none' is not a widget role

### PR DESCRIPTION
**Pull request by nmgokhale**
Fixes: #1402 

Followed same pattern as [carbon tooltip](https://v10.carbondesignsystem.com/components/tooltip/code/). 
- For toolbar - Added `aria-labelledby` attribute to the `tooltip-trigger` element.
- For common-properties tooltips - Added `aria-labelledby, aria-expanded, aria-controls` attributes to the `tooltip-trigger` element. 

Follwing [ARIA example](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/tooltip_role#examples) -
Added `role="tooltip", id` attributes to actual tooltip content having `className="common-canvas-tooltip"`.

**Before -**
Screenshot 2023-03-16 at 5 04 14 PM

**After -**
Screenshot 2023-03-16 at 5 17 11 PM

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: canvas_modules/common-canvas/src/common-properties/controls/textarea/textarea.jsx [hunk 1]
```diff
@@ -98,7 +98,7 @@ class TextareaControl extends React.Component {
 
 		let display = textArea;
 		if (this.props.tableControl) {
-			const tooltipId = uuid4() + "-tooltip-column-" + this.props.propertyId.toString();
+			const tooltipId = uuid4() + "-tooltip-column-" + getPropertyIdOfTableControlString(this.props.propertyId);
```
Comment (nmgokhale) [inline canvas_modules/common-canvas/src/common-properties/controls/textarea/textarea.jsx:101]: Before there was `[Object Object]` in tooltipId  -
Screenshot 2023-03-16 at 5 28 31 PM

After -
Screenshot 2023-03-16 at 5 33 56 PM

File: canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx [hunk 2]
```diff
@@ -245,7 +245,7 @@ class VirtualizedTable extends React.Component {
 			? null
 			: (<div className="properties-vt-info-icon-tip">
 				<Tooltip
-					id={`properties-tooltip-${columnData.headerLabel}-info`}
+					id={`properties-tooltip-${columnData.headerLabel.split(" ").join("")}-info`}
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx:248]: This is the same as `string.replace(" ", "")`?

File: canvas_modules/common-canvas/src/common-properties/util/property-utils.js [hunk 3]
```diff
@@ -573,6 +573,18 @@ function convertValueDataTypes(currentParameters, paramDefs) {
 	return convertedCurrentParameters;
 }
 
+// Return string value for {name: string, row: integer, col: integer}
+function getPropertyIdOfTableControlString(propertyId) {
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/common-properties/util/property-utils.js:577]: Note that we do support nesting in properties, for example in structuretable_paramDef:
image

Take a look at `propertiesController.convertNestedPropertyId()`

File: canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx (hunk 2 above)
Comment (matthoward366) [reply canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx:248]: This doesn't seem like a good way of creating ids.  There could be other chars in the header label that cause issues and there is no guarantee these are unique in the DOM.  Can we generate a guid for each header instead?  We wouldn't want to generate it on each rerender though.

File: canvas_modules/common-canvas/src/common-properties/controls/textarea/textarea.jsx (hunk 1 above)
Comment (matthoward366) [reply canvas_modules/common-canvas/src/common-properties/controls/textarea/textarea.jsx:101]: Are we getting new uuid each time we rerender?  That doesn't seem like a good idea.  Should we move the uuid to the constructor so it's only generated once for this class?

File: canvas_modules/common-canvas/src/common-properties/controls/textfield/textfield.jsx [hunk 4]
```diff
@@ -109,7 +109,7 @@ class TextfieldControl extends React.Component {
 
 		let display = textInput;
 		if (this.props.tableControl) {
-			const tooltipId = uuid4() + "-tooltip-column-" + this.props.propertyId.toString();
+			const tooltipId = uuid4() + "-tooltip-column-" + getPropertyIdOfTableControlString(this.props.propertyId);
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/controls/textfield/textfield.jsx:112]: Same comment as above

File: canvas_modules/common-canvas/src/common-properties/util/property-utils.js (hunk 3 above)
Comment (matthoward366) [reply canvas_modules/common-canvas/src/common-properties/util/property-utils.js:577]: Take a look at https://github.com/elyra-ai/canvas/blob/main/canvas_modules/common-canvas/src/common-properties/util/control-utils.js#L25.  I think we already have a method that does this.

As long as these id's are unique I don't think we really need to care about them being nested.

File: canvas_modules/common-canvas/src/common-properties/controls/textarea/textarea.jsx (hunk 1 above)
Comment (nmgokhale) [reply canvas_modules/common-canvas/src/common-properties/controls/textarea/textarea.jsx:101]: Yes there's a new `uuid` for each rerender and all table cells have a different `uuid`. I will generate `uuid` once in constructor and use it for all tooltipIds.

File: canvas_modules/common-canvas/src/tooltip/tooltip.jsx [hunk 5]
```diff
@@ -406,7 +411,14 @@ class ToolTip extends React.Component {
 		if (tooltipContent || link) {
 			tooltip = (
 				<Portal>
-					<div data-id={this.props.id} className={tipClass} aria-hidden={!this.state.isTooltipVisible} direction={this.props.direction}>
+					<div
+						role="tooltip"
+						id={`${this.uuid}-${this.props.id}`}
```
Comment (nmgokhale) [inline canvas_modules/common-canvas/src/tooltip/tooltip.jsx:416]: Generating `UUID` in tooltip. Prefix it to tooltip id. Consuming component can pass a generic `id` in props. UUID makes sure tooltipId is unique for all.

File: canvas_modules/common-canvas/src/tooltip/tooltip.jsx [hunk 6]
```diff
@@ -368,13 +373,13 @@ class ToolTip extends React.Component {
 
 		if ((typeof this.props.tip) === "string") {
 			tooltipContent = (
-				<span id="tooltipContainer">
+				<span id={`${this.uuid}-tooltipContainer`}>
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/tooltip/tooltip.jsx:376]: For the `id` do we actually need to set `id`?  I'm curious if it's actually needed.  This seems like it should be a class or not set.

File: canvas_modules/common-canvas/src/tooltip/tooltip.jsx (hunk 5 above)
Comment (nmgokhale) [reply canvas_modules/common-canvas/src/tooltip/tooltip.jsx:416]: We need an `id` for `aria-labelledby` attribute to work. 
We can remove `data-id` now that `id` will be unique after using uuid. But many jest tests will break after removing data-id, so I kept both id and data-id.

Comment (matthoward366) [reply canvas_modules/common-canvas/src/tooltip/tooltip.jsx:416]: Ah, ok.

**Review by matthoward366 — APPROVED** [review]
Looks good

**Review by caritaou — APPROVED** [review]

**Review by tomlyn — APPROVED** [review]

**Pull request merged**